code/controllers/subsystem/blackbox.dm

```diff
@@ -0,0 +1,307 @@
+
+
+/**
+  * Helper to find and return a feeedback datum
+  *
+  * Pass in a feedback datum key and key_type to do a lookup.
+  * It will create the feedback datum if it doesnt exist
+  *
+  * Arguments:
+  * * key - Key of the variable to lookup
+  * * key_type - Type of feedback to be recorded if the feedback datum cant be found
+  */
+/datum/controller/subsystem/blackbox/proc/find_feedback_datum(key, key_type)
+	for(var/datum/feedback_variable/FV in feedback)
+		if(FV.key == key)
+			return FV
+
+	var/datum/feedback_variable/FV = new(key, key_type)
+	feedback += FV
+	return FV
+
+/**
+  * Main feedback recording proc
+  *
+  * This is the bulk of this subsystem and is in charge of creating and using the variables.
+  * See .github/USING_FEEDBACK_DATA.md for instructions
+  * Note that feedback is not recorded to the DB during this function. That happens at round end.
+  *
+  * Arguments:
+  * * key_type - Type of key. Either "text", "amount", "tally", "nested tally", "associative"
+  * * key - Key of the data to be used (EG: "admin_verb")
+  * * increment - If using "amount", how much to increment why
+  * * data - The actual data to logged
+  * * overwrite - Do we want to overwrite the existing key
+  */
+/datum/controller/subsystem/blackbox/proc/record_feedback(key_type, key, increment, data, overwrite)
+	if(sealed || !key_type || !istext(key) || !isnum(increment || !data))
+		return
+	var/datum/feedback_variable/FV = find_feedback_datum(key, key_type)
+	switch(key_type)
+		if("text")
+			if(!istext(data))
+				return
+			if(!islist(FV.json["data"]))
+				FV.json["data"] = list()
+			if(overwrite)
+				FV.json["data"] = data
+			else
+				FV.json["data"] |= data
+		if("amount")
+			FV.json["data"] += increment
+		if("tally")
+			if(!islist(FV.json["data"]))
+				FV.json["data"] = list()
+			FV.json["data"]["[data]"] += increment
+		if("nested tally")
+			if(!islist(data))
+				return
+			if(!islist(FV.json["data"]))
+				FV.json["data"] = list()
+			FV.json["data"] = record_feedback_recurse_list(FV.json["data"], data, increment)
+		if("associative")
+			if(!islist(data))
+				return
+			if(!islist(FV.json["data"]))
+				FV.json["data"] = list()
+			var/pos = length(FV.json["data"]) + 1
+			FV.json["data"]["[pos]"] = list()
+			for(var/i in data)
+				FV.json["data"]["[pos]"]["[i]"] = "[data[i]]"
+
+/**
+  * Recursive list recorder
+  *
+  * Used by the above proc for nested tallies
+  *
+  * Arguments:
+  * * L - List to use
+  * * key_list - List of keys to add
+  * * increment - How much to increase by
+  * * depth - Depth to use
+  */
+/datum/controller/subsystem/blackbox/proc/record_feedback_recurse_list(list/L, list/key_list, increment, depth = 1)
+	if(depth == key_list.len)
+		if(L.Find(key_list[depth]))
+			L["[key_list[depth]]"] += increment
+		else
+			var/list/LFI = list(key_list[depth] = increment)
+			L += LFI
+	else
+		if(!L.Find(key_list[depth]))
+			var/list/LGD = list(key_list[depth] = list())
+			L += LGD
```

Comment:
the mnemonics of `LFI` and `LGD` are entirely arcane to me. List.. First.. Increment and List... Global.. Data..?